### PR TITLE
Enable Gitpod workspaces with pre-installed dependencies

### PR DESCRIPTION
**Pull request by pattacini**
This PR drops in the `.gitpod.yml` file pointing to the Docker image that provides the dependencies required to run the experiment.

This is, in turn, the Docker file used to generate the image: https://github.com/pattacini/dockerfiles/blob/master/visual-tactile-localization/Dockerfile.

Here's how to enable a VNC browser-based access:

noVNC

This is a valid first step in the direction outlined here: https://spectrum.chat/icub/general/get-yarp-enabled-workspaces-on-the-cloud-with-a-single-click~53ba1286-4d46-4402-94f5-4aa321f03fb5?m=MTU1NTYwODYzNjA3Ng==

### CODE REVIEW

**Comment by pattacini** [comment]
@xEnVrE and I are still looking into why the repo is not compilable within the Docker container due to some errors spotted on the use of `Assimp` library.

This PR can be merged anyway.

**Review by xEnVrE — APPROVED** [review]
LGTM! Thanks :tada:

**Comment by xEnVrE** [comment]
In the end, I was able to solve the problems with `assimp` and `glm` but I needed to add a **temporary** fix in `CMakeLists.txt` files of modules `object-tracking` and `object-tracking-viewer`. (Part of )The root of these problems seems to be in `YCM` and I am investigating in order to solve the issues in the proper way. The temporary fix is already in `master`.

Unfortunately there are other issues that have to be solved.

#### VNC crashes
When I open the `yarpdataplayer` something very wrong happens when I try to click on `File` in the menu of the application, i.e., the `vnc` session crashes.

 After this I tried to restart the vnc session by typing `start-vnc-session.sh` in the gitpod terminal, I opened again the vnc viewer and I found this situation
vnc_error

Once the error window is closed, everything is smooth and the server does not crash anymore. 

The same happens, for example, when clicking on an item of the menu of the `yarpmanager`. However, it only happens the first time the `gitpod` is run. After the first crash, and after the vnc server is restarted, the issue does not present anymore.

I don't know if we already have a repository for gitpod/docker related stuff. If we have one, I can open a separate issue to report this problem.

#### Yarpview does not open
To be precise, `yarpview` opens correctly if launched from the `gitpod` terminal. 

On the other hand, if launched from the `yarpmanager` or from a terminal, within the vnc viewer, it does not open. When launched from the terminal these errors are printed:

vnc_yarpview

Again, I think it is better if I open a separate issue for this problem.

#### Method glfwCreateWindow from library GLFW fails
The method `glfwCreateWindow` is used within the library `SuperimposeMeshLib` during the initialization of the rendering facilities. This probably depends on the hardware configuration, i.e. GPU, of the virtual machine offered by `gitpod` and need to be investigated.

As a consequence, the `object-tracking` module cannot be started within this environment at the moment.

I will open an issue also for this in my workbook.

**Comment by pattacini** [comment]
Well done with the debug @xEnVrE 👍 

The script `start-vnc-session.sh` is responsible for launching all the required X11 components: see https://github.com/pattacini/dockerfiles/blob/708613b722459f6302ea5443c692448049779452/visual-tactile-localization/Dockerfile#L134-L139.

Thereby, when you launch it manually, after the first crash, the service `xcfe4-panel` will be running twice, bringing up the error message you saw. I didn't experience similar problems when I was fiddling with the xml applications of smaller tutorials. Also, I didn't try to launch `yarpview` from within a `terminator` console. You might want to file separate issues on https://github.com/pattacini/dockerfiles.

Finally, I'm afraid I've underestimated the requirements of GPU that seems needed to let `SuperimposeMeshLib` run smoothly. As of now, Gitpod doesn't sport GPU in its environments. Maybe, we could get around such a limitation asking `SuperimposeMeshLib` to give up with GPU to rely only on CPU. Is this possible?

**Comment by pattacini** [comment]
Already files an issue for the problem with `yarpview`.

**Comment by pattacini** [comment]
The problem with `yarpview` should be solved: please, wait for the Docker image to be ready.

**Comment by pattacini** [comment]
Importantly, the patch above should also fix the crash happening while clicking on `File`.

**Comment by pattacini** [comment]
❓ What is left is the question "are we able to run `SuperimposeMeshLib` w/o GPU?"

**Comment by claudiofantacci** [comment]
> ❓ What is left is the question "are we able to run `SuperimposeMeshLib` w/o GPU?"

You don't really need a powerful GPU, you just need a GPU that exposes OpenGL version 3.3 or greater. The problem is that, on some virtual machine, OpenGL support is fixed to version 2.*.

**Comment by pattacini** [comment]
Hi @claudiofantacci 
I'm afraid we don't have GPU at all on Gitpod 😞 
Are we able to run all the ensemble on CPU only? We have many CPU's instead...

**Comment by claudiofantacci** [comment]
Hey @pattacini, unfortunately, we cannot. The library uses render facilities provided by OpenGL that can be run only on GPU 😢

**Comment by pattacini** [comment]
Ok damn'...
@xEnVrE @claudiofantacci is there any subset of the experiment that could be eventually run on Gitpod, meaning w/o GPU? I'm thinking of disabling the superimposition of the mesh. We won't have the same visual effect but at least the filter should function.

**Comment by pattacini** [comment]
Maybe this can be of help.
https://stackoverflow.com/questions/7310885/opengl-without-a-graphics-card

**Comment by pattacini** [comment]
@claudiofantacci if you could gimme quick feedback on http://tipsonubuntu.com/2019/02/08/install-mesa-graphics-driver-18-3-3-ubuntu-18-04 I'll then try it out on Gitpod.

**Comment by xEnVrE** [comment]
> is there any subset of the experiment that could be eventually run on Gitpod, meaning w/o GPU? I'm thinking of disabling the superimposition of the mesh. We won't have the same visual effect but at least the filter should be functioning well.

Unfortunately I am using the facilities of the library for some computation related to the update of the bounding box. I think I am gonna search for gitpods enabled repositories using opengl. Maybe we are lucky. I can also test for `mesa` :smiley:

**Comment by pattacini** [comment]
@xEnVrE the [image with `mesa` is being built][1] on Docker Hub, hence wait for it.
Concerning GPU-enabled Gitpod environments, my concrete fear is that you won't find any.

[1]: https://cloud.docker.com/repository/registry-1.docker.io/pattacini/visual-tactile-localization/builds/a79d07b3-f81d-453d-8eff-646fc4ccbe02

**Comment by claudiofantacci** [comment]
Isn't mesa present in the standard repo? I didn't know that without a GPU it could run OpenGL code on CPU. We can try to test it 😄 I'm now wondering whether this is actually the standard behavior of the VM. They just use mesa and if the version is too old then also the OpenGL support is also as well.

Sorry, I could not reply faster. Anyway it is worth a shot 👍

**Comment by pattacini** [comment]
> ✔️ Importantly, the patch above should also fix the crash happening while clicking on `File`.

@xEnVrE this is actually a ❌ since I've tested it over. Don't know yet what causes the problem.

@claudiofantacci this `mesa` is not apparently provided from standard repos since it needs a PPA.

**Comment by claudiofantacci** [comment]
I just read the FAQ and got the point of the situation. Yes, for Ubuntu there are some ppa with precompiled Mesa 3D. Notice that the latest version supports up to version 3.1. Not sure whether the library will work or not. I don’t honestly remember whether the version is a suggestion for minimum required version that can roll back to a compatible older one or not.

**Comment by pattacini** [comment]
> > ✔️ Importantly, the patch above should also fix the crash happening while clicking on `File`.
>
> @xEnVrE this is actually a ❌ since I've tested it over. Don't know yet what causes the problem.

https://github.com/pattacini/dockerfiles/commit/a27a622c5141151f1f287289b16f1dead363dda9 should hopefully fix the problem. In essence, `x11vnc` is very old and we need the latest package instead 🤞

**Comment by pattacini** [comment]
@xEnVrE preliminary tests indicate that the latest `x11vnc` seems to have fixed the initial crash 🚀

**Comment by xEnVrE** [comment]
I also tried it before and I confirm no more crashes for now :rocket: 

I have a question. I tried to look [here](https://cloud.docker.com/repository/docker/pattacini/visual-tactile-localization/builds) to see whether the last image with `mesa` was done. However, it seems that the last `build` is failed. When i try to access the gitpod environment, i.e. with [this](https://gitpod.io/#https://github.com/robotology-playground/visual-tactile-localization), I get the error
```
An error occurred while pulling image for container: workspace
```

Is it normal and is it related to the failed build?

**Comment by pattacini** [comment]
The build of the image with `mesa` is still pending 'cause I had to tinker with some flags: it will be ready reasonably tomorrow by then.

For the rest, I've just clicked on the link and it's working: I've got my Gitpod instance over the repo.
Try again, sometimes it might be stuck.

**Comment by pattacini** [comment]
The image Is ready 👍

**Comment by pattacini** [comment]
@xEnVrE no way, just tested unsuccessfully.
I also realized that `mesa-utils` was missing: just added it up. Building again, although I suspect it won't be enough.

**Comment by xEnVrE** [comment]
First of all I tried to confirm that we have `OpenGL`. I extracted the executable `glxinfo` from the `.deb` of `mesa-utils` and by typing `glxinfo | grep "OpenGL"` we have
```
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: llvmpipe (LLVM 8.0, 256 bits)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 19.0.2
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.1 Mesa 19.0.2
OpenGL shading language version string: 1.40
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.0 Mesa 19.0.2
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
OpenGL ES profile extensions:
```

However, the creation of the `GLFW` window still fails. By using the GLFW error reporting facilities, the errors that prevent the creation of the window are the following
```
GLFW TEST Start setting up...
[INFO]Calling glfwInit() 
Error: X11: RandR gamma ramp support seems broken
Error: Linux: Failed to watch for joystick connections in /dev/input: No such file or directory
Error: Linux: Failed to open joystick device directory /dev/input: No such file or directory
[INFO]Calling glfwCreateWindow 
Error: GLX: Failed to create context: BadValue (integer parameter out of range for operation)
[ERROR]Failed to create GLFWwindow 
```

where, according to some sources,`Error: X11: RandR gamma ramp support seems broken` may depend on the fact that headless rendering, as done inside `SuperimposeMesh` is supported only starting from `GLFW 3.3` while we have `GLFW 3.2`. However, `glfwInit()` does not fail.

Instead we need to investigate on the second error `GLX: Failed to create context: BadValue (integer parameter out of range for operation)`.

**Comment by pattacini** [comment]
Good 👍 
`mesa-utils` is now in the queue; however, I've stopped the building 'cause the latest `yarp/devel` is bugged. Waiting for the fix before starting over.

**Comment by xEnVrE** [comment]
The fact is that `glxgears`, a OpenGL rendering of a simple arrangement of three rotating gears,  is running smoothly, hence the rendering stuff is somehow working. Probably is something related to the fact that we use `vnc`. 

Maybe iit is worth checking the configuration of [this](https://github.com/thewtex/docker-opengl/blob/master/Dockerfile) docker, called `docker-opengl`, to see if we are missing something.

**Comment by pattacini** [comment]
@xEnVrE having `vnc` interfering with OpenGL would be weird since it's only a wrapper to let the user see what's inside the video buffer.

A quick thing to try to rule this out:
- kill `x11vnc` and `novnc` services and just let `Xvfb` run
- repeat your test

**Comment by pattacini** [comment]
@xEnVrE from the list of packages installed in `docker-opengl` I can spot `libgl1-mesa-dri`.
Does this get installed in our image? I presume you're working locally with Docker since in Gitpod you wouldn't be able to try out different installations.

**Comment by xEnVrE** [comment]
In the end I found out what the problem was.

It seems that the error
```
Error: GLX: Failed to create context: BadValue (integer parameter out of range for operation)
``` 
was caused by [this](https://github.com/robotology/superimpose-mesh-lib/blob/feaab46e95d675770496c87a190b46abc4f3d991/src/SuperimposeMesh/src/SICAD.cpp#L105) line of code
```
 glfwWindowHint(GLFW_CONTEXT_RELEASE_BEHAVIOR, GLFW_RELEASE_BEHAVIOR_NONE);
```
, that tells `GLFW` what to to do with the "pipeline" when the context is released.

I am not really sure whether commenting out this hint is harmful or not, however by doing it `GLFW` is now able to initialize the context properly and `SuperimposeMesh` is now working properly. 

This is the output of the test [test_sicad_model](https://github.com/robotology/superimpose-mesh-lib/tree/master/test/test_sicad_model) (this test uses the rendering facilities used within the module `object-tracking` - and also `hand-tracking`) :tada:

test_sil

No modifications are required to the docker file since, in the end, enabling `OSMesa`, i.e. off screen rendering without even the need for a running `X` server, in the building options of `GLFW` and `GLEW` was useless. The context was initialized properly but then there were errors in the attachment of the `color` and `depth` buffers to the context.

I will test everything together soon.

**Comment by pattacini** [comment]
That's super amazing! 🚀

**Comment by claudiofantacci** [comment]
Thanks @xenvre. Commenting out that line of code is potentially harmful as I remember I had to set it to have the pipeline working properly across different OSs (and maybe there are also some performance issue involved). However, as we updated SICAD several time, I’m no longer sure it is really required and maybe de default behaviour is now just fine. We need to test everything properly!

**Comment by pattacini** [comment]
@xEnVrE but if we need to apply a change to `SuperImposeMesh` then we need to regenerate the docker, unless this change can be done into this repo. Let me know.

**Comment by xEnVrE** [comment]
> We need to test everything properly!

I totally agree with this!

> but if we need to apply a change to SuperImposeMesh then we need to regenerate the docker

For now I'll try to test it locally (having some issues with noVNC in my local docker installation).

> unless this change can be done into this repo. Let me know.

I will discuss f2f with @claudiofantacci about that line of code an I'll let you know.

**Comment by xEnVrE** [comment]
Working ok in my local docker container

Screenshot from 2019-04-28 18-34-31

although frame rate is somehow reduced

Screenshot from 2019-04-28 18-35-47

**Comment by pattacini** [comment]
Fantastic news!

**Comment by claudiofantacci** [comment]
### tl;dr
We can probably fix this problem by changing [this](https://github.com/robotology/superimpose-mesh-lib/blob/feaab46e95d675770496c87a190b46abc4f3d991/src/SuperimposeMesh/src/SICAD.cpp#L105) line of code in
```cpp
if(GLEW_KHR_context_flush_control)
    glfwWindowHint(GLFW_CONTEXT_RELEASE_BEHAVIOR, GLFW_RELEASE_BEHAVIOR_NONE);
```

### Full explanation
It might be that the Mesa 3D does not implement the [`KHR_context_flush_control`](https://www.khronos.org/registry/OpenGL/extensions/KHR/KHR_context_flush_control.txt) extension and by default the OpenGL pipeline is flushed when the OpenGL context is no longer current (that may happen several times using specific `SuperimposeMesh` calls. Since we are using `GLEW`, me should be able to check whether the extension is present on a machine by checking the boolean variable `GLEW_<extension name>`, i.e. by checking whether `GLEW_KHR_context_flush_control` exists or not (see associated issue [here](https://github.com/nigels-com/glew/issues/194)).

**Comment by pattacini** [comment]
@claudiofantacci superb! 🎉
I'me flooding this thread with very encouraging adverbs, but for me using Gitpod for submissions is really transformative. `Code Ocean` doesn't allow for such a granularity.

**Comment by claudiofantacci** [comment]
> @claudiofantacci superb! 🎉
> I'me flooding this thread with very encouraging adverbs, but for me using Gitpod for submissions is really transformative. `Code Ocean` doesn't allow for such a granularity.

Even though I have GitPod enabled since forever, I have never used it and I now feel I'm losing a big chunk of expertise here. I'll try to get up to your point asap and maybe get some KB fro @xEnVrE during one of the next update meeting.

**Comment by xEnVrE** [comment]
Thanks @claudiofantacci.

I tried but, unfortunately, it seems that all the variables `GLEW_<extensionname>` are populated only after a call to `glewInit()`. However, if I am not wrong, we can call `glewInit()` only after the creation of the context using `glfwCreateWindow()`, and at that point all the hints, like `glfwWindowHint(GLFW_CONTEXT_RELEASE_BEHAVIOR, GLFW_RELEASE_BEHAVIOR_NONE)`, have been already set.

The bad notice, however, is that the implementation of Mesa that we are using inside the docker image **does** support the context flush control, i.e. `GLEW_KHR_context_flush_control == GL_TRUE` after a call to `glewInit()`. The same can also be verified using `glewinfo`

```
glewinfo | grep "flush_control"
```
Output on docker installation:
```
GL_KHR_context_flush_control:                                  OK 
GLX_ARB_context_flush_control:                                 OK 
```
Error when creating a window using `GLFW` (when `glfwWindowHint(GLFW_CONTEXT_RELEASE_BEHAVIOR, GLFW_RELEASE_BEHAVIOR_NONE)` is called)
```
Error: GLX: Failed to create context: BadValue (integer parameter out of range for operation)
```

**Comment by claudiofantacci** [comment]
This is really strange. By commenting out that specific line of code, the default behaviour for flushing the pipeline is assumed, otherwise is changed.

After calling `glfwCreateWindow()` we should try to get the last OpenGL error following [this guide.](https://www.glfw.org/docs/latest/intro_guide.html#error_handling). We have a range of error we can get after creating a window (see [here](https://www.glfw.org/docs/latest/group__window.html#ga5c336fddf2cbb5b92f65f10fb6043344)), and I suspect the error we have is either [`GLFW_API_UNAVAILABLE`](https://www.glfw.org/docs/latest/group__errors.html#ga196e125ef261d94184e2b55c05762f14) or [`GLFW_INVALID_VALUE`](https://www.glfw.org/docs/latest/group__errors.html#gaaf2ef9aa8202c2b82ac2d921e554c687).

Anyway, we can test `Superimpose` without that line and see what happen.

**Comment by claudiofantacci** [comment]
@xEnVrE what version of GLFW are you using? Please note that the feature is implemented in GLFW starting from version `3.1`.

**Comment by xEnVrE** [comment]
The error

```
Error: GLX: Failed to create context: BadValue (integer parameter out of range for operation)
```
was obtained using the `GLFW` error reporting system based on callbacks that prints a human-readable description described in the guide you were referring to. So probably we have a `GLFW_INVALID_VALUE`.

**Comment by xEnVrE** [comment]
> @xEnVrE what version of GLFW are you using?

I am using `GLFW 3.2.1-1`.

**Comment by claudiofantacci** [comment]
Since `glxinfo` reports OpenGL version `3.1`, can you please try to change
```cpp
glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
```
in [this](https://github.com/robotology/superimpose-mesh-lib/blob/feaab46e95d675770496c87a190b46abc4f3d991/src/SuperimposeMesh/src/SICAD.cpp#L100) with
```cpp
glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 1);
```

**Comment by xEnVrE** [comment]
Hi @claudiofantacci,

I tried as you suggested. I had to set `GLFW_OPENGL_CORE_PROFILE` to `GLFW_OPENGL_ANY_PROFILE` in [this](https://github.com/robotology/superimpose-mesh-lib/blob/feaab46e95d675770496c87a190b46abc4f3d991/src/SuperimposeMesh/src/SICAD.cpp#L101) since it is [mandatory](https://www.glfw.org/docs/latest/window_guide.html#GLFW_OPENGL_PROFILE_hint) for OpenGL version below 3.2.

However, the result is the same.

> I'm afraid I can't do more without trying it myself and I don't know exactly how to set the whole thing up.

Maybe we can talk about this tomorrow. However, I think that , for now, we can also decide to patch `SuperimposeMesh` only within the docker image since everything is working by commenting out that line.

**Comment by pattacini** [comment]
> Maybe we can talk about this tomorrow. However, I think that , for now, we can also decide to patch `SuperimposeMesh` only within the docker image since everything is working by commenting out that line.

It would be sufficient to create an _ad-hoc_ branch that I will checkout in the Docker file.

**Comment by xEnVrE** [comment]
### tl; dr
In the end the reason of the problem is `Xvfb`, the X virtual framebuffer that we are using in the docker image. We should probably use something more recent like [Xdummy](https://xpra.org/trac/wiki/Xdummy) or [xserver-xorg-video-dummy](https://techoverflow.net/2019/02/23/how-to-run-x-server-using-xserver-xorg-video-dummy-driver-on-ubuntu/).

### Full explanation
According to the source code of the the `Xvfb` server, when a `GLX` context is created, all the attributes, i.e. what in `GLFW` are called "window hints", are processed in a for cycle

```
#include "glxext.h"
(...)
for (i = 0; i < req->numAttribs; i++) {
    switch (attribs[i * 2]) {
       case GLX_CONTEXT_MAJOR_VERSION_ARB:
            major_version = attribs[2 * i + 1];
            break;
(...)

#ifdef GLX_CONTEXT_RELEASE_BEHAVIOR_ARB
        case GLX_CONTEXT_RELEASE_BEHAVIOR_ARB:
            flush = attribs[2 * i + 1];
            if (flush != GLX_CONTEXT_RELEASE_BEHAVIOR_NONE_ARB
                && flush != GLX_CONTEXT_RELEASE_BEHAVIOR_FLUSH_ARB)
                return BadValue;
            break;
#endif

        default:
              return BadValue;
}
```
The part associated to the context release/flush depends on the variable `GLX_CONTEXT_RELEASE_BEHAVIOR_ARB` that is defined in the `OpenGL` file `GL/glxext.h`. Unfortunately, the authors of the code decided, in their local copy of `glxext.h`, that
```
/* doing #include <GL/glx.h> & #include <GL/glxext.h> could cause problems        
 * with overlapping definitions, so let's use the easy way
 */
(...)
```
where the "easy way" involves taking a subset of the `GLX_*` variables and excluding, for example, `GLX_CONTEXT_RELEASE_BEHAVIOR_ARB`, `GLX_CONTEXT_RELEASE_BEHAVIOR_NONE_ARB` and `GLX_CONTEXT_RELEASE_BEHAVIOR_FLUSH_ARB`.

Hence, when a `GLFW` window hint involving the context release is issued, there is a specific value of `i` such that `attribs[i * 2] == GLX_CONTEXT_RELEASE_BEHAVIOR_ARB`. Since that part of code is not compiled, the default `return BadValue` is issued resulting in

```
Error: GLX: Failed to create context: BadValue (integer parameter out of range for operation)
```

**Comment by pattacini** [comment]
@xEnVrE nice catch 👍 
I will have some time probably tomorrow to inspect the use of more recent frame buffers, the ones you suggested.

**Comment by claudiofantacci** [comment]
@xEnVrE, now we're talking. It was unlikely that there was another reason other than a missing extension definition behind our problem. Good catch indeed and let's see whether by changing the video driver.

**Comment by xEnVrE** [comment]
Probably also all the other frame buffers have the same problem since that code is not specific to `Xvfb` but is part of the general `xorg` server code (and shared by `Xvfb`). I think that we never experienced issues because with real GPUs Linux uses `DRI` to communicate directly with the hardware and skips the buggy code.

**Comment by xEnVrE** [comment]
Good news, I opened an [issue](https://gitlab.freedesktop.org/xorg/xserver/issues/684) for that and a user created a [PR](https://gitlab.freedesktop.org/xorg/xserver/merge_requests/183) to fix the issue.

Then, we can just build the latest `xorg` on the docker image and everything should be fine!

**Comment by pattacini** [comment]
Well done @xEnVrE 

Anyway, I'm in the process of replacing `Xvfb` with `xserver-xorg-video-dummy`.

Let me know when the PR gets merged so that I'll generate the docker images again.

**Comment by pattacini** [comment]
It works! Finally! 🚀
